core/src/main/java/org/apache/iceberg/MergingSnapshotProducer.java

```diff
@@ -309,6 +327,31 @@ protected void validateDataFilesExist(TableMetadata base, Long startingSnapshotI
         VALIDATE_DATA_FILES_EXIST_SKIP_DELETE_OPERATIONS :
         VALIDATE_DATA_FILES_EXIST_OPERATIONS;
 
+    Pair<List<ManifestFile>, Set<Long>> history =
+        validationHistory(base, startingSnapshotId, matchingOperations, ManifestContent.DATA);
+    List<ManifestFile> manifests = history.first();
+    Set<Long> newSnapshots = history.second();
+
+    ManifestGroup matchingDeletesGroup = new ManifestGroup(ops.io(), manifests, ImmutableList.of())
+        .filterManifestEntries(entry -> entry.status() != ManifestEntry.Status.ADDED &&
+            newSnapshots.contains(entry.snapshotId()) && requiredDataFiles.contains(entry.file().path()))
+        .specsById(base.specsById())
+        .ignoreExisting();
+
+    try (CloseableIterator<ManifestEntry<DataFile>> deletes = matchingDeletesGroup.entries().iterator()) {
+      if (deletes.hasNext()) {
+        throw new ValidationException("Cannot commit, missing data files: %s",
```

Comment:
"Cannot commit, deletes have been added to the table which refer to data files which no longer exist"?

I think I go that right here ... It would help me a lot if the message here was a bit more explanatory about what went wrong and why.

Reply:
This is an existing error message. We can fix it, but I'd prefer to do it in a separate commit so that we don't have changes to more test cases than necessary. I also plan to fix the initial validation in `RewriteFiles` later for the same reason. Is it okay to leave this as-is?